fdbserver/VersionedBTree.actor.cpp

```diff
@@ -3423,13 +3423,32 @@ class DWALPager final : public IPager2 {
 		}
 
 		if (freeNewID) {
-			debug_printf("DWALPager(%s) remapCleanup freeNew %s\n", self->filename.c_str(), p.toString().c_str());
-			self->freeUnmappedPage(p.newPageID, 0);
+			debug_printf("DWALPager(%s) remapCleanup freeNew %s %s\n",
+			             self->filename.c_str(),
+			             p.toString().c_str(),
+			             toString(self->getLastCommittedVersion()).c_str());
+
+			// newID must be freed at the latest committed version to avoid a read race between caching and non-caching
+			// readers. It is possible that there are readers of newID in flight right now that either
+			//  - Did not read through the page cache
+			//  - Did read through the page cache but there was no entry for the page at the time, so one was created
```

Comment:
Active readers at a version  _do_ hold up oldestRetainedVersion so it can't advance past that version, but that does not enable freeing the page as of the next commit version.  

The problem is with a read of the newer page version, the target of the remap.  Such reads can happen at versions all the way up to latestCommittedVersion and would be redirected and read the remap target page.  If they did so without going through the cache, then the read could complete after a new write of the page if it was freed for immediate reuse.  Example: 

- At v10, page X is mapped to Y. 
- oldestRetainedVersion is v18
- latestCommittedVersion is v25
- A read at v25 reads page X and is redirected to Y.  This read is not going through the cache, so it's reading page Y from disk directly, and this disk read is slow, it has not yet completed.
- The remap entry of  X -> Y at v10 iis removed by remap cleanup, so Y data is copied to X and then Y is freed as of the end of the commit
- The commit ends
- The next commit allocates a page and gets Y from the free list and writes to it, and this write completes.
- The earlier dispatched read of Y from disk which did not read through the cache finally completes, and it sees the updated data which is wrong.

Reply:
To be very precise, the version at which the target page can be freed at is the latest version at which the page is currently being actively read as of the time the remap entry is removed.  But tracking this would be expensive, so it's easier to just free the page as of latestCommitVersion.

Reply:
Thanks for the example.

In case `freeNewID=true`, there's always the next remap entry ("R R |" and "R F |"). Instead of delay until `lastCommitedVersion`, can we delay only till the version at the next remap entry?

Reply:
Yes I think so, delay free at `lastCommittedVersion` should only be required for
`R |`
but for all other cases of `freeNewID == true` the version of the next change should be sufficient, which are
```
R R |
R F |
R | R
R | F
```

I would rather not make that change in this PR though in case there is some edge we are missing.

Reply:
Nice catch btw, do you want to make a PR for it?

Reply:
ok, would do later.